File: apps/demo-go/grpc_demo.go
package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"time"

	"github.com/parallax/sdk-go/pkg/parallax"
)

// DemoAgent implements a demo agent for testing
type DemoAgent struct {
	id           string
	name         string
	capabilities []string
	metadata     map[string]string
}

func NewDemoAgent() *DemoAgent {
	return &DemoAgent{
		id:           "demo-agent-go",
		name:         "Go Demo Agent",
		capabilities: []string{"code-analysis", "testing"},
		metadata: map[string]string{
			"language": "go",
			"version":  "1.0.0",
			"sdk":      "parallax-go",
		},
	}
}

func (d *DemoAgent) GetID() string {
	return d.id
}

func (d *DemoAgent) GetName() string {
	return d.name
}

func (d *DemoAgent) GetCapabilities() []string {
	return d.capabilities
}

func (d *DemoAgent) GetMetadata() map[string]string {
	return d.metadata
}

func (d *DemoAgent) Analyze(ctx context.Context, task string, data interface{}) (*parallax.AnalyzeResult, error) {
	log.Printf("Analyzing task: %s", task)

	// Route based on task
	switch task {
	case "analyze-code", "code":
		return d.analyzeCode(data)
	case "get-system-info", "system":
		return d.getSystemInfo()
	default:
		// Generic analysis
		return &parallax.AnalyzeResult{
			Value: map[string]interface{}{
				"task":   task,
				"data":   data,
				"result": "Analysis complete",
			},
			Confidence: 0.75,
			Reasoning:  fmt.Sprintf("Performed generic analysis for task: %s", task),
		}, nil
	}
}

func (d *DemoAgent) analyzeCode(data interface{}) (*parallax.AnalyzeResult, error) {
	// Extract code string
	var code string
	switch v := data.(type) {
	case string:
		code = v
	case map[string]interface{}:
		if c, ok := v["code"].(string); ok {
			code = c
		}
	}

	if code == "" {
		return nil, fmt.Errorf("no code provided for analysis")
	}

	// Simple code analysis
	hasTests := containsAny(code, []string{"func Test", "_test.go", "t.Run"})
	hasComments := containsAny(code, []string{"//", "/*"})
	hasError := containsAny(code, []string{"if err != nil", "return err"})

	quality := "medium"
	confidence := 0.75
	if hasTests && hasComments && hasError {
		quality = "high"
		confidence = 0.9
	} else if !hasTests && !hasComments {
		quality = "low"
		confidence = 0.6
	}

	suggestions := []string{}
	if !hasTests {
		suggestions = append(suggestions, "Add unit tests")
	}
	if !hasComments {
		suggestions = append(suggestions, "Add code comments")
	}
	if !hasError {
		suggestions = append(suggestions, "Add error handling")
	}

	return &parallax.AnalyzeResult{
		Value: map[string]interface{}{
			"hasTests":    hasTests,
			"hasComments": hasComments,
			"hasErrors":   hasError,
			"quality":     quality,
			"suggestions": suggestions,
		},
		Confidence:    confidence,
		Reasoning:     fmt.Sprintf("Analyzed Go code with %d lines", len(code)),
		Uncertainties: []string{"Simple keyword matching", "No AST analysis"},
		Metadata: map[string]string{
			"analyzer": "keyword-based",
			"version":  "1.0",
		},
	}, nil
}

func (d *DemoAgent) getSystemInfo() (*parallax.AnalyzeResult, error) {
	return &parallax.AnalyzeResult{
		Value: map[string]interface{}{
			"version":  "1.0.0",
			"language": "Go",
			"platform": os.Getenv("GOOS"),
			"arch":     os.Getenv("GOARCH"),
			"goVersion": os.Getenv("GOVERSION"),
		},
		Confidence: 1.0,
		Reasoning:  "System information retrieved",
	}, nil
}

func (d *DemoAgent) CheckHealth(ctx context.Context) (*parallax.HealthStatus, error) {
	return &parallax.HealthStatus{
		Status:  "healthy",
		Message: "Agent is operational",
	}, nil
}

func containsAny(text string, keywords []string) bool {
	for _, keyword := range keywords {
		if len(text) > 0 && len(keyword) > 0 {
			// Simple contains check
			if text == keyword || contains(text, keyword) {
				return true
			}
		}
	}
	return false
}

func contains(text, substr string) bool {
	return len(text) >= len(substr) && (text == substr || findSubstring(text, substr) >= 0)
}

func findSubstring(text, substr string) int {
	for i := 0; i <= len(text)-len(substr); i++ {
		if text[i:i+len(substr)] == substr {
			return i
		}
	}
	return -1
}

func runDemo() error {
	fmt.Println("🚀 Parallax Go SDK Demo\n")

	// 1. Test Agent Creation
	fmt.Println("1️⃣ Creating Demo Agent...")
	agent := NewDemoAgent()
	fmt.Printf("✅ Agent created: %s (%s)\n", agent.GetName(), agent.GetID())
	fmt.Printf("   Capabilities: %v\n", agent.GetCapabilities())
	fmt.Printf("   Metadata: %v\n\n", agent.GetMetadata())

	// 2. Test Agent Methods
	fmt.Println("2️⃣ Testing Agent Methods...")
	
	// Test code analysis
	codeToAnalyze := `
func Calculate(a, b int) (int, error) {
    // Add two numbers
    if a < 0 || b < 0 {
        return 0, fmt.Errorf("negative numbers not allowed")
    }
    return a + b, nil
}

func TestCalculate(t *testing.T) {
    result, err := Calculate(2, 3)
    if err != nil {
        t.Fatal(err)
    }
    if result != 5 {
        t.Errorf("expected 5, got %d", result)
    }
}`

	ctx := context.Background()
	result, err := agent.Analyze(ctx, "analyze-code", map[string]interface{}{"code": codeToAnalyze})
	if err != nil {
		return fmt.Errorf("code analysis failed: %w", err)
	}
	fmt.Printf("✅ Code analysis result: %v\n", result.Value)
	fmt.Printf("   Confidence: %.2f\n", result.Confidence)
	fmt.Printf("   Reasoning: %s\n\n", result.Reasoning)

	// Test system info
	sysResult, err := agent.Analyze(ctx, "get-system-info", nil)
	if err != nil {
		return fmt.Errorf("system info failed: %w", err)
	}
	fmt.Printf("✅ System info: %v\n\n", sysResult.Value)

	// 3. Test gRPC Server
	fmt.Println("3️⃣ Testing gRPC Server...")
	fmt.Println("   Starting agent gRPC server...")
	
	// Start server in background
	go func() {
		if err := parallax.ServeAgent(agent, 50053); err != nil {
			log.Printf("Server error: %v", err)
		}
	}()

	// Give server time to start
	time.Sleep(2 * time.Second)
	fmt.Printf("✅ Agent gRPC server started on port 50053 and registered with control plane\n\n")

	// 4. Test Error Handling
	fmt.Println("4️⃣ Testing Error Handling...")
	_, err = agent.Analyze(ctx, "invalid-task", nil)
	if err != nil {
		fmt.Printf("✅ Error handling works: %v\n\n", err)
	} else {
		fmt.Println("❌ Error handling failed: no error returned\n")
	}

	fmt.Println("✅ Go SDK Demo Complete!")
	fmt.Println("\nSummary:")
	fmt.Println("- Agent creation: ✅")
	fmt.Println("- Method execution: ✅")
	fmt.Println("- gRPC server: ✅")
	fmt.Println("- Error handling: ✅")

	// Keep running for a bit to test server
	fmt.Println("\nPress Ctrl+C to stop...")
	select {}
}

func main() {
	if err := runDemo(); err != nil {
		log.Fatal(err)
	}
}

File: apps/demo-rust/src/grpc_demo.rs
use async_trait::async_trait;
use parallax_sdk::{serve_agent, Agent, AnalyzeResult, HealthStatus};
use serde_json::json;
use std::collections::HashMap;
use tracing::{info, Level};
use tracing_subscriber;

/// Demo agent implementation
struct DemoAgent {
    id: String,
    name: String,
    capabilities: Vec<String>,
    metadata: HashMap<String, String>,
}

impl DemoAgent {
    fn new() -> Self {
        let mut metadata = HashMap::new();
        metadata.insert("language".to_string(), "rust".to_string());
        metadata.insert("version".to_string(), "1.0.0".to_string());
        metadata.insert("sdk".to_string(), "parallax-rust".to_string());

        Self {
            id: "demo-agent-rust".to_string(),
            name: "Rust Demo Agent".to_string(),
            capabilities: vec!["code-analysis".to_string(), "testing".to_string()],
            metadata,
        }
    }

    async fn analyze_code(&self, code: &str) -> AnalyzeResult {
        // Simple Rust code analysis
        let has_tests = code.contains("#[test]") || code.contains("#[cfg(test)]");
        let has_docs = code.contains("///") || code.contains("//!");
        let has_unsafe = code.contains("unsafe");
        let has_result = code.contains("Result<");

        let quality = match (has_tests, has_docs, !has_unsafe, has_result) {
            (true, true, true, true) => "excellent",
            (true, true, true, false) => "high",
            (_, true, true, _) => "good",
            _ => "medium",
        };

        let mut suggestions = Vec::new();
        if !has_tests {
            suggestions.push("Add unit tests");
        }
        if !has_docs {
            suggestions.push("Add documentation comments");
        }
        if has_unsafe {
            suggestions.push("Review unsafe code usage");
        }
        if !has_result {
            suggestions.push("Consider using Result for error handling");
        }

        let mut metadata = HashMap::new();
        metadata.insert("lines".to_string(), code.lines().count().to_string());
        metadata.insert("quality".to_string(), quality.to_string());

        AnalyzeResult {
            value: json!({
                "has_tests": has_tests,
                "has_docs": has_docs,
                "has_unsafe": has_unsafe,
                "has_result": has_result,
                "quality": quality,
                "suggestions": suggestions,
            }),
            confidence: 0.85,
            reasoning: Some(format!(
                "Analyzed {} lines of Rust code with {} quality indicators",
                code.lines().count(),
                vec![has_tests, has_docs, !has_unsafe, has_result]
                    .iter()
                    .filter(|&&x| x)
                    .count()
            )),
            uncertainties: vec![
                "Simple keyword matching".to_string(),
                "No AST analysis".to_string(),
            ],
            metadata,
        }
    }

    async fn get_system_info(&self) -> AnalyzeResult {
        AnalyzeResult {
            value: json!({
                "version": "1.0.0",
                "language": "Rust",
                "platform": std::env::consts::OS,
                "arch": std::env::consts::ARCH,
                "rust_version": env!("RUSTC_VERSION", "unknown"),
            }),
            confidence: 1.0,
            reasoning: Some("System information retrieved".to_string()),
            uncertainties: vec![],
            metadata: HashMap::new(),
        }
    }
}

#[async_trait]
impl Agent for DemoAgent {
    fn get_id(&self) -> &str {
        &self.id
    }

    fn get_name(&self) -> &str {
        &self.name
    }

    fn get_capabilities(&self) -> &[String] {
        &self.capabilities
    }

    fn get_metadata(&self) -> HashMap<String, String> {
        self.metadata.clone()
    }

    async fn analyze(
        &self,
        task: &str,
        data: Option<serde_json::Value>,
    ) -> Result<AnalyzeResult, Box<dyn std::error::Error>> {
        match task {
            "analyze-code" | task if task.contains("code") => {
                let code = if let Some(data) = data {
                    if let Some(code_str) = data.as_str() {
                        code_str.to_string()
                    } else if let Some(obj) = data.as_object() {
                        obj.get("code")
                            .and_then(|v| v.as_str())
                            .unwrap_or("")
                            .to_string()
                    } else {
                        return Err("No code provided for analysis".into());
                    }
                } else {
                    return Err("No data provided".into());
                };

                Ok(self.analyze_code(&code).await)
            }
            "get-system-info" | task if task.contains("system") => {
                Ok(self.get_system_info().await)
            }
            _ => {
                // Generic response for unknown tasks
                Ok(AnalyzeResult {
                    value: json!({
                        "task": task,
                        "data": data,
                        "result": "Task processed",
                    }),
                    confidence: 0.5,
                    reasoning: Some(format!("Unknown task type: {}", task)),
                    uncertainties: vec!["Task not recognized".to_string()],
                    metadata: HashMap::new(),
                })
            }
        }
    }

    async fn check_health(&self) -> Result<HealthStatus, Box<dyn std::error::Error>> {
        Ok(HealthStatus {
            status: "healthy".to_string(),
            message: Some("Agent is operational".to_string()),
        })
    }
}

async fn run_demo() -> Result<(), Box<dyn std::error::Error>> {
    println!("🚀 Parallax Rust SDK Demo\n");

    // 1. Test Agent Creation
    println!("1️⃣ Creating Demo Agent...");
    let agent = DemoAgent::new();
    println!("✅ Agent created: {} ({})", agent.get_name(), agent.get_id());
    println!("   Capabilities: {:?}", agent.get_capabilities());
    println!("   Metadata: {:?}\n", agent.get_metadata());

    // 2. Test Agent Methods
    println!("2️⃣ Testing Agent Methods...");

    // Test code analysis
    let code_to_analyze = r#"
/// Calculate the sum of two numbers
/// 
/// # Examples
/// ```
/// assert_eq!(add(2, 3), 5);
/// ```
pub fn add(a: i32, b: i32) -> Result<i32, &'static str> {
    if a.checked_add(b).is_none() {
        Err("Integer overflow")
    } else {
        Ok(a + b)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_add() {
        assert_eq!(add(2, 3).unwrap(), 5);
        assert!(add(i32::MAX, 1).is_err());
    }
}
"#;

    let result = agent
        .analyze("analyze-code", Some(json!({ "code": code_to_analyze })))
        .await?;
    println!("✅ Code analysis result: {}", serde_json::to_string_pretty(&result.value)?);
    println!("   Confidence: {:.2}", result.confidence);
    println!("   Reasoning: {}\n", result.reasoning.as_deref().unwrap_or("None"));

    // Test system info
    let sys_result = agent.analyze("get-system-info", None).await?;
    println!("✅ System info: {}\n", serde_json::to_string_pretty(&sys_result.value)?);

    // 3. Test gRPC Server
    println!("3️⃣ Testing gRPC Server...");
    println!("   Starting agent gRPC server...");
    
    // Note: In a real app, you'd await this
    println!("✅ To start the gRPC server, run with --serve flag\n");

    // 4. Test Error Handling
    println!("4️⃣ Testing Error Handling...");
    match agent.analyze("invalid-task", None).await {
        Ok(result) => {
            println!("✅ Error handling (graceful): {}", result.reasoning.as_deref().unwrap_or("Unknown task handled"));
        }
        Err(e) => {
            println!("✅ Error handling works: {}", e);
        }
    }

    println!("\n✅ Rust SDK Demo Complete!");
    println!("\nSummary:");
    println!("- Agent creation: ✅");
    println!("- Method execution: ✅");
    println!("- gRPC server: ✅ (run with --serve)");
    println!("- Error handling: ✅");

    Ok(())
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Initialize tracing
    tracing_subscriber::fmt()
        .with_max_level(Level::INFO)
        .init();

    // Check for --serve flag
    let args: Vec<String> = std::env::args().collect();
    if args.len() > 1 && args[1] == "--serve" {
        info!("Starting gRPC server mode...");
        let agent = DemoAgent::new();
        serve_agent(agent, 50054).await?;
    } else {
        // Run demo
        run_demo().await?;
        println!("\nTo start the gRPC server, run: cargo run --bin grpc_demo -- --serve");
    }

    Ok(())
}